pages/blog/socketio-automatic-docs.md

```diff
@@ -0,0 +1,553 @@
+
+def check_game_status(board: List[List[str]]) -> Union[str, None]:
+    # Check for a win or draw
+    for i in range(3):
+        if board[i] == ['X', 'X', 'X']:
+            return 'X'
+        if board[i] == ['O', 'O', 'O']:
+            return 'O'
+    if all(board[i][i] == 'X' for i in range(3)):
+        return 'X'
+    if all(board[i][i] == 'O' for i in range(3)):
+        return 'O'
+    if all(board[i][j] != '' for i in range(3) for j in range(3)):
+        return 'draw'
+    return None
+
+
+if __name__ == '__main__':
+    socketio.run(app, debug=True)
+
+```
+
+It's not quite correct (e.g. data in make_move will be passed as a dictionary not a Pydantic model), but it's good enough for our purposes.
+
+### Generate Mermaid diagram
+Now let's ask ChatGPT to generate a Mermaid diagram for us as well, so we can get a better illustration of our server:
+
+>- create mermaid diagram for previous example
+
+I had to alter the given diagram a bit, but here is the final result:
+
+```mermaid
+sequenceDiagram
+    participant Client
+    participant SocketIOServer
+
+    Client ->> SocketIOServer: create_game
+    SocketIOServer -->> Client: game_created(game_id)
+
+    Client ->> SocketIOServer: make_move(game_id, x, y)
+    alt if_not_won_or_drawn
+    SocketIOServer -->> Client: move_made(board, turn)
+    end
+
+    alt if_won
+    SocketIOServer -->> Client: game_won(winner)
+    end
+
+    alt if_drawn
+    SocketIOServer -->> Client: game_drawn
+    end
+```
+
+
+### Add SIO-AsyncAPI to the server
+
+Now let's imagine that this is our old server written in Flask-SocketIO and we want to add SIO-AsyncAPI to it. Here is how we would do it:
+
+1. Install SIO-AsyncAPI `pip install sio-asyncapi`
+2. Change import statements
+```python
+# instead of `from flask_socketio import SocketIO`
+from sio_asyncapi import AsyncAPISocketIO as SocketIO
+```
+3. Add additional argument to `AsyncAPISocketIO` constructor
+```python
+socketio = SocketIO(
+    app,
+    validate=True,
+    generate_docs=True,
+    version="1.0.0",
+    title="Tick-Tack-Toe API",
+    description="Tick-Tack-Toe Game API",
+    server_url="http://localhost:5000",
+    server_name="TICK_TACK_TOE_BACKEND",
+)
+```
+5. Tell `@socketio.on` decorator to get models from typehint.
+> Note: you can also pass `request_model` and `response_model` arguments to `@socketio.on` decorator instead of using typehints.
+```python
+@socketio.on('make_move', get_from_typehint=True)
+```
+Now type annotations will be used to generate AsyncAPI specification and validate incoming and outgoing messages. Note that the return value from a function is not data sent by `emit` function but rather `acknowledge` value that is received by the client.
+
+6. Add `on_emit` decorator to register/document a SocketIO emit event. Since we are not defining `emit` and only calling it as a function, we need to tell SIO-AsyncAPI what to expect when `emit` is called. E.g
```

Comment:
```suggestion
6. Add an `on_emit` decorator to register/document a SocketIO emit event. Since we are not defining `emit` and only calling it a function, we need to tell SIO-AsyncAPI what to expect when `emit` is called. E.g.
```

Reply:
I rephrased it a bit here. I think it was not clear when I said "as a function"